lib/kube/proxy/streamproto/proto.go

```diff
@@ -0,0 +1,269 @@
+
+	if !isClient && !ok {
+		return nil, trace.BadParameter("handshake must be either client or server handshake")
+	}
+
+	if isClient {
+		ty, data, err := conn.ReadMessage()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if ty != websocket.TextMessage {
+			return nil, trace.Errorf("expected websocket control message, got %v", ty)
+		}
+
+		var msg metaMessage
+		if err := utils.FastUnmarshal(data, &msg); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if msg.ServerHandshake == nil {
+			return nil, trace.Errorf("expected websocket server handshake, got %v", msg)
+		}
+
+		s.MFARequired = msg.ServerHandshake.MFARequired
+		handshakeMsg := metaMessage{ClientHandshake: &clientHandshake}
+		dataClientHandshake, err := utils.FastMarshal(handshakeMsg)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := conn.WriteMessage(websocket.TextMessage, dataClientHandshake); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	} else {
+		handshakeMsg := metaMessage{ServerHandshake: &serverHandshake}
+		dataServerHandshake, err := utils.FastMarshal(handshakeMsg)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := conn.WriteMessage(websocket.TextMessage, dataServerHandshake); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		ty, data, err := conn.ReadMessage()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if ty != websocket.TextMessage {
+			return nil, trace.Errorf("expected websocket control message, got %v", ty)
+		}
+
+		var msg metaMessage
+		if err := utils.FastUnmarshal(data, &msg); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if msg.ClientHandshake == nil {
+			return nil, trace.Errorf("expected websocket client handshake, got %v", msg)
+		}
+
+		s.Mode = msg.ClientHandshake.Mode
+	}
+
+	go s.readTask()
+	return s, nil
+}
+
+func (s *SessionStream) readTask() {
+	for {
+		terminated := false
+		defer s.closedC.Do(func() { close(s.CloseC) })
+		ty, data, err := s.conn.ReadMessage()
+		if err != nil {
+			return
+		}
+
+		if ty == websocket.BinaryMessage {
+			s.in <- data
+		}
+
+		if ty == websocket.TextMessage {
+			var msg metaMessage
+			if err := utils.FastUnmarshal(data, &msg); err != nil {
+				return
+			}
+
+			if msg.Resize != nil {
+				s.resizeQueue <- msg.Resize
+			}
+
+			if msg.ForceTerminate && !terminated {
+				terminated = true
+				close(s.forceTerminate)
+			}
+		}
+
+		if ty == websocket.CloseMessage {
+			s.conn.Close()
+			s.closed = true
+			return
+		}
+	}
+}
+
+func (s *SessionStream) Read(p []byte) (int, error) {
+	if len(s.currentIn) == 0 {
+		select {
+		case s.currentIn = <-s.in:
+		case <-s.CloseC:
+			return 0, io.EOF
+		}
+	}
+
+	n := copy(p, s.currentIn)
+	s.currentIn = s.currentIn[n:]
+	return n, nil
+}
+
+func (s *SessionStream) Write(data []byte) (int, error) {
+	s.writeSync.Lock()
+	defer s.writeSync.Unlock()
+
+	return len(data), s.conn.WriteMessage(websocket.BinaryMessage, data)
```

Comment:
If `WriteMessage` returns an error I would consider returning `0, err` to comply with `io.Writer`.